src/qcodes/dataset/database_extract_runs.py

```diff
@@ -168,3 +178,190 @@ def _extract_single_dataset_into_db(
+                if exp_id not in processed_experiments:
+                    exp_attrs = get_experiment_attributes_by_exp_id(source_conn, exp_id)
+
+                    with atomic(target_conn) as target_conn:
+                        target_exp_id = _create_exp_if_needed(
+                            target_conn,
+                            exp_attrs["name"],
+                            exp_attrs["sample_name"],
+                            exp_attrs["format_string"],
+                            exp_attrs["start_time"],
+                            exp_attrs["end_time"],
+                        )
+
+                    processed_experiments[exp_id] = target_exp_id
+                    log.info(
+                        f"Created experiment '{exp_attrs['name']}' with ID {target_exp_id} in target database"
+                    )
+                else:
+                    target_exp_id = processed_experiments[exp_id]
+
+                # Try to export dataset to NetCDF and create metadata-only version
+                status = _process_single_dataset(
+                    dataset, source_conn, target_conn, export_path, target_exp_id
+                )
+                result_status[run_id] = status
+
+            except Exception as e:
+                log.exception(f"Failed to process dataset {run_id}")
+                result_status[run_id] = "failed"
+
+    log.info("Exporting complete.")
+    return result_status
+
+
+def _process_single_dataset(
+    dataset: DataSetProtocol,
+    source_conn: AtomicConnection,
+    target_conn: AtomicConnection,
+    export_path: Path,
+    target_exp_id: int,
+) -> Literal["exported", "copied_as_is", "failed"]:
+    """
+    Export a dataset to NetCDF and add its metadata
+    to target database file, or, if it fails, copy directily
+    into the target database file.
+
+    Returns:
+        Status string indicating what was done with the dataset
+
+    """
+    run_id = dataset.run_id
+
+    netcdf_export_path = None
+
+    existing_netcdf_path = dataset.export_info.export_paths.get("nc")
+    if existing_netcdf_path is not None:
+        existing_path = Path(existing_netcdf_path)
+        # Check if the existing export path matches the desired export path
+        if existing_path.exists() and existing_path.parent == export_path:
+            log.debug(
+                f"Dataset {run_id} already exported to NetCDF at {existing_netcdf_path}"
+            )
+            netcdf_export_path = existing_netcdf_path
+        else:
+            log.info(
+                f"Dataset {run_id} was exported to different location, re-exporting to {export_path}"
+            )
+    else:
+        log.debug(f"Attempting to export dataset {run_id} to NetCDF")
+
+    if netcdf_export_path is None:
+        try:
+            dataset.export("netcdf", path=export_path)
+            netcdf_export_path = dataset.export_info.export_paths.get("nc")
+            assert netcdf_export_path is not None
```

Comment:
We should not use assert outside test code so this should be rewritten to explicitly raise a more descriptive error